### PR TITLE
commit hooks: new permission/notification system

### PR DESCRIPTION
**Pull request by matthewrmshin**
The new system is based on simple permission and notification modes at
repository, project and/or branch level, where owners can be specified.

Logic is unified in the Perl module `FCM::Admin::Commit`.
New configuration settings (provisional) are documented at the bottom
of the new module.

### CODE REVIEW

File: lib/FCM/Admin/Commit.pm [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+        );
+        my $project = $layout->get_project();
+        my $branch = $layout->get_branch();
+
+        # Notify branch owner
+        if (    $commit_conf->get_notification_modes()->{'branch'}
+            &&  $layout->is_branch()
+            &&  defined($branch)
+        ) {
+            my $branch_path = join(q{/}, grep {$_} ($project, $branch));
```
Comment (benfitzpatrick) [inline lib/FCM/Admin/Commit.pm]: Why not `join(q{/}, $project, $branch)`?

Comment (benfitzpatrick) [inline lib/FCM/Admin/Commit.pm]: (because $project may not be defined..., retracted)

Comment (matthewrmshin) [inline lib/FCM/Admin/Commit.pm]: Yes.

File: lib/FCM/Admin/Commit.pm [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+            }
+            next LINE;
+        }
+        # Project owner can do anything to paths at or under the project
+        if (    $commit_conf->get_permission_modes()->{'project'}
+            &&  defined($project)
+        ) {
+            if (!_perm_ok($commit_conf, $author, $project)) {
+                # Addition of paths as part of a branch creation is OK
+                # Permission denied for everything else under the project
```
Comment (benfitzpatrick) [inline lib/FCM/Admin/Commit.pm]: Why are they allowed to make a branch at all?

Comment (matthewrmshin) [inline lib/FCM/Admin/Commit.pm]: Comment should be *project creation*. I'll change that.

File: lib/FCM/Admin/Commit.pm [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+        if ($owner) {
+            return $owner;
+        }
+    }
+    return;
+}
+
+# Get and return the youngest revision of $repos
+memoize('_get_youngest_rev');
+sub _get_youngest_rev {
```
Comment (benfitzpatrick) [inline lib/FCM/Admin/Commit.pm]: OK to memoize this?

File: lib/FCM/Admin/Commit.pm (hunk 2 above)
Comment (matthewrmshin) [inline lib/FCM/Admin/Commit.pm]: Sorry, I meant *tag creation*, etc.

File: lib/FCM/Admin/Commit.pm (hunk 3 above)
Comment (matthewrmshin) [inline lib/FCM/Admin/Commit.pm]: Should be OK for the purpose of the commit hook.

File: t/fcm-install-svn-hook/00-basic.t [hunk 4]
```diff
@@ -131,14 +131,14 @@ NAME='svnperms-conf' run_tests
 file_cmp "$TEST_KEY-ls-svnperms.conf" \
     'svn-repos/foo/hooks/svnperms.conf' 'svn-import/svnperms.conf'
 
-# New install, single repository, with commit.conf
-TEST_KEY="$TEST_KEY_BASE-commit.conf"
+# New install, single repository, with commit.cfg
+TEST_KEY="$TEST_KEY_BASE-commit.cfg"
 {
     echo 'notify-branch-owner'
     echo 'verify-branch-owner'
```
Comment (benfitzpatrick) [inline t/fcm-install-svn-hook/00-basic.t]: Why keep these?

File: lib/FCM/Admin/Commit.pm [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+=item owner[C<project>]=C<USER1> ...
+=item owner[C<project/branches/dev/Share/whatever>]=C<USER1> ...
+
+Specify the user IDs of the owners of a project (or a shared topic branches
+of). The owners of a project can make changes to the trunk and any branches in
+the project. The owners of a named shared branch can make changes to the shared
+branch. If C<project> is in C<notification-modes>, project owners will be
+informed of all changes that are within a project but outside of its topic
+branches. If C<branch> is in C<notification-modes>, branch owners will be
+informed of all changes that are within the branch.
```
Comment (benfitzpatrick) [inline lib/FCM/Admin/Commit.pm]: This needs a bit saying that users won't be informed of their own changes.

**Comment by benfitzpatrick** [comment]
Some comments, otherwise looks good.

**Comment by matthewrmshin** [comment]
Added a post-commit subscriber functionality.

N.B. branch re-based.

File: lib/FCM/Admin/Commit.pm [hunk 6]
```diff
@@ -0,0 +1,571 @@
...
+                else {
+                    push(@bads, [$line]);
+                }
+            }
+            next LINE;
+        }
+        # Repository owner can do anything
+        if ($commit_conf->get_permission_modes()->{'repository'}) {
+            if (!_perm_ok($commit_conf, $author)) {
+                # An author who is not a project owner can add paths under the
```
Comment (benfitzpatrick) [inline lib/FCM/Admin/Commit.pm]: `repository owner`

File: lib/FCM/Admin/Commit.pm [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+                # project as part of a branch creation, project creation, etc.
+                # "$line_of_unknown_path{$path}" may be deleted if it is a part
+                # of a valid branch creation, project creation, etc. by an
+                # author who is not a project owner.
```
Comment (benfitzpatrick) [inline lib/FCM/Admin/Commit.pm]: Can I make a new project if I include a new branch as part of the transaction?

**Comment by benfitzpatrick** [comment]
Some comments, otherwise looks fine to me.

File: lib/FCM/Admin/Commit.pm (hunk 7 above)
Comment (matthewrmshin) [inline lib/FCM/Admin/Commit.pm]: Logic + test added to prevent this.

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by matthewrmshin** [comment]
@scwhitehouse please sanity check the logic, and pick on anything that looks suspicious. Documentation of the proposed configuration can be found at the bottom of the `FCM::Admin::Commit` module. I'll add more documentation once this change is agreed and finalised.

**Comment by benfitzpatrick** [comment]
(branch has conflicts)

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by scwhitehouse** [comment]
After Matt's very kindly sat down and explained this, I'm happy with this as a reviewer.

**Comment by matthewrmshin** [comment]
Branch rebased.

**Comment by matthewrmshin** [comment]
We have now decided to not pursue this change in the foreseeable future. Now that the system is quite mature, it is best to avoid breaking it. We'll reopen this if there is demand.

**Pull request closed without merging**